Editorial suggestions from Benjamin Kaduk (recovery) (#4685)
quicwg/base-drafts: draft-ietf-quic-recovery.md

a packet number).  The encryption level indicates the packet number space, as
described in {{QUIC-TRANSPORT}}.  Packet numbers never repeat within a packet
number space for the lifetime of a connection.  Packet numbers are sent in
monotonically increasing order within a space, preventing ambiguity.  It is

> permitted for some packet numbers to never be used, leaving intentional gaps.

Reviewer: Do we now also have to acknowledge the fact that the loss recovery algorithms assume the absence of gaps?

Reviewer 2: The algorithm does not assume the absence of gaps -- Section 6.1 says: `The packet was sent kPacketThreshold packets before an acknowledged packet`. The pseudo-code does assume it though, and where it does, it is stated in a comment (see A.10).

All in all, I think this addition is safe.

Author: My expectations matched what Jana describes.

## Estimating smoothed_rtt and rttvar {#smoothed-rtt}

smoothed_rtt is an exponentially-weighted moving average of an endpoint's RTT

> samples, and rttvar represents the variation in the RTT samples, estimated using a

Author: (if it's "estimated", then it's not "is")

Reviewer: It is if the definition of the term is an representation of an estimate, which is what it is.

Author: I don't think the comma is appropriate if the claim is that rttvar is defined to be the estimated variation of the RTT sample.

Reviewer 3: I can change this to "and rttvar is the mean variation in the RTT samples." if that's preferable?

## Estimating smoothed_rtt and rttvar {#smoothed-rtt}

smoothed_rtt is an exponentially-weighted moving average of an endpoint's RTT

> samples, and rttvar represents the variation in the RTT samples, estimated using
> a mean variation.

Reviewer: ```suggestion
samples, and rttvar estimates the variation in the RTT samples using a mean
variation.
```

The peer might report acknowledgment delays that are larger than the peer's
max_ack_delay during the handshake (Section 13.2.1 of {{QUIC-TRANSPORT}}). To
account for this, the endpoint SHOULD ignore max_ack_delay until the handshake

> is confirmed (as defined in Section 4.1.2 of {{QUIC-TLS}}). When they occur, these large

Reviewer 3: ```suggestion
is confirmed, as defined in Section 4.1.2 of {{QUIC-TLS}}. When they occur, these large
```

of the round-trip time and for the correct packet across packet number spaces.

When a PTO timer expires, the PTO backoff MUST be increased, resulting in the

> PTO period being set to twice its current value, in each packet number space. The PTO backoff factor is reset

Author: Okay, sorry for the confusion.  I was just keying off of:

>   Even when there are ack-
   eliciting packets in-flight in multiple packet number spaces, the
   exponential increase in probe timeout occurs across all spaces to
   prevent excess load on the network.  For example, a timeout in the
   Initial packet number space doubles the length of the timeout in the
   Handshake packet number space.

> PTO period being set to twice its current value, in each packet number space. The PTO backoff factor is reset

Reviewer 3: Yeah, in this case across all is not quite the same as for each, but I agree the difference is subtle.

## Discarding Keys and Packet State {#discarding-packets}

When Initial and Handshake packet protection keys are discarded
(see Section 4.9 of {{QUIC-TLS}}), all packets that were sent with those keys

> can no longer be acknowledged because their acknowledgments cannot be processed
> anymore. The sender MUST discard all recovery state associated with those packets

Reviewer: ```suggestion
can no longer be acknowledged because their acknowledgments cannot be processed.
The sender MUST discard all recovery state associated with those packets
```

## Discarding Keys and Packet State {#discarding-packets}

> When packet protection keys are discarded from an entire encryption level (see Section 4.9 of {{QUIC-TLS}}),

Author: I believe that we refer to the 1-RTT key update procedure as "discarding" the old generation of keys.  My understanding is that the loss of recovery state described here does not apply to 1-RTT key update.

Reviewer 4: ```suggestion
When Initial and Handshake packet protection keys are discarded (see Section 4.9 of {{QUIC-TLS}}),
```

before Initial packets, early 0-RTT packets will be declared lost, but that
is expected to be infrequent.

> It is expected that keys are discarded after all packets encrypted with them would

Reviewer 4: I'm not sure if this is correct. You could receive HANDSHAKE_DONE without having received all ACKs for Handshake packets.

Author: Hmm.  I considered s/all/any/ but am not sure that helps.  The best idea I have right now (just before bedtime...) is to add to the next sentence that Handshake keys are discarded when the handshake is confirmed.

congestion SHOULD consider packets sent across packet number spaces. A sender
that does not have state for all packet number spaces or an implementation that
cannot compare send times across packet number spaces MAY use state for just the

> packet number space that was acknowledged.  This might result in erroneously
> declaring persistent congestion when including all packet number spaces would
> not have declared persistent connection, but will not lead to a failure to
> detect persistent congestion.

Reviewer: ```suggestion
packet number space that was acknowledged.  This might result in erroneously
declaring persistent congestion.
```

> packet number space that was acknowledged.  This might result in erroneously
> declaring persistent congestion when including all packet number spaces would
> not have declared persistent connection, but will not lead to a failure to
> detect persistent congestion.

Reviewer 2: ```suggestion
packet number space that was acknowledged.  This might result in erroneously
declaring persistent congestion, but it will not lead to a failure to detect persistent
congestion. 
```

: The largest packet number acknowledged in the packet number space so far.

loss_time\[kPacketNumberSpace]:

> : The earliest time at which an additional packet in that packet number space will be

Reviewer 4: What was wrong with the original sentence?

Author: "next" could be ambiguous about applying to the time axis or the packet number axis.

Reviewer 4: As packet numbers are used sequentially, those axis should be effectively equivalent.

Reviewer 3: +1 to Marten's comment, reverting most of it, but changing will to can, since timers can wake up late.

  // min_rtt ignores acknowledgment delay.
  min_rtt = min(min_rtt, latest_rtt)
  // Limit ack_delay by max_ack_delay after handshake

>   // confirmation. Note that ack_delay is clamped to 0 for

Author: ack_delay is an input to this function, and I didn't remember any external constraint that the provided function argument would be zero in this case.

    loss_detection_timer.cancel()
    return

>   // Determine which PN space to arm PTO for.

Author: stale comment?

Reviewer 3: Let's just remove this comment.

Author: Do you want me to be making updates to the PR, or can you take care of anything that's needed?

Reviewer 3: I'm happy to do it.

Author: Thank you!

  // Enter recovery period.
  congestion_recovery_start_time = now()

>   ssthresh = congestion_window * (1 - kLossReductionFactor)

Author: I tried to triple check this, but there is not much to compare against.
If the loss reduction factor is supposed to indicate the amount of **reduction**, we need this change.  When the value is 0.5 it doesn't matter, of course...

Reviewer: The factor is multiplicative in cubic (0.7 there), so 1-B is likely to be problematic.

Author: So we should change the definition of kLossReductionFactor to be something like "scaling factor applied to reduce the congestion window when loss is detected"?
